zcash_keys/src/keys.rs

```diff
@@ -678,42 +708,69 @@ impl UnifiedFullViewingKey {
             sapling,
             #[cfg(feature = "orchard")]
             orchard,
-            unknown,
+            unknown_data,
+            expiry_height,
+            expiry_time,
+            unknown_metadata,
         })
     }
 
     /// Returns the string encoding of this `UnifiedFullViewingKey` for the given network.
     pub fn encode<P: consensus::Parameters>(&self, params: &P) -> String {
-        let items = std::iter::empty().chain(self.unknown.iter().map(|(typecode, data)| {
-            unified::Fvk::Unknown {
-                typecode: *typecode,
-                data: data.clone(),
-            }
-        }));
+        let data_items =
+            std::iter::empty().chain(self.unknown_data.iter().map(|(typecode, data)| {
+                unified::Fvk::Unknown {
+                    typecode: *typecode,
+                    data: data.clone(),
+                }
+            }));
         #[cfg(feature = "orchard")]
-        let items = items.chain(
+        let data_items = data_items.chain(
             self.orchard
                 .as_ref()
                 .map(|fvk| fvk.to_bytes())
                 .map(unified::Fvk::Orchard),
         );
         #[cfg(feature = "sapling")]
-        let items = items.chain(
+        let data_items = data_items.chain(
             self.sapling
                 .as_ref()
                 .map(|dfvk| dfvk.to_bytes())
                 .map(unified::Fvk::Sapling),
         );
         #[cfg(feature = "transparent-inputs")]
-        let items = items.chain(
+        let data_items = data_items.chain(
             self.transparent
                 .as_ref()
                 .map(|tfvk| tfvk.serialize().try_into().unwrap())
                 .map(unified::Fvk::P2pkh),
         );
 
-        let ufvk = unified::Ufvk::try_from_items(items.collect())
-            .expect("UnifiedFullViewingKey should only be constructed safely");
+        let meta_items = std::iter::empty()
+            .chain(self.unknown_metadata.iter().map(|(typecode, data)| {
+                unified::MetadataItem::Unknown {
+                    typecode: *typecode,
+                    data: data.clone(),
+                }
+            }))
+            .chain(
+                self.expiry_height
+                    .map(|h| unified::MetadataItem::ExpiryHeight(u32::from(h))),
+            )
+            .chain(self.expiry_time.map(unified::MetadataItem::ExpiryTime));
+
+        let ufvk = unified::Ufvk::try_from_items(
+            if self.expiry_height().is_some() || self.expiry_time().is_some() {
+                Revision::R1
+            } else {
+                Revision::R0
```

Comment:
This causes the construction to fail for a transparent-only UFVK without expiry height or time. That's fine, but we should enforce this policy consistently.

Reply:
Fixed in 078b4ffa4577c8cb0e1a325b6a1b2b512c20a18a